### ISSUE TITLE
post_url generates urls without baseurl

### ISSUE THREAD

**Issue by Coder-RG** [jekyll/jekyll#8887]
## My Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Operating System | MacOS Monterey v12.0.1 |
| `jekyll`         | 4.2.1 |
| `github-pages`   | 219 |

---

`_config.yml` excerpt:
```yaml
baseurl: /blog
permalink: none
```
none-> /:categories/:title:output_ext

## Expected Behaviour
post_url should generate complete URL that work without any extra input.

Example:
Consider two posts, DATE-article1.md and DATE-article2.md. The expected URL is `/blog/category/article1.md`.

## Current Behavior
Currently, the generated URL don't contain site.baseurl value. This breaks those links.

Example:
Consider two posts, DATE-article1.md and DATE-article2.md. The generated URL is `/category/article1.md`.

**NOTE:** If this is not a bug, then please consider this a feature request.

## Code Sample

Considering two articles, DATE-article1.md and DATE-article2.md.

- Without Explicit baseurl
```
[Article 1]({% post_url DATE-article1.md %})
```
>ERROR `/category/article1.html' not found.

- With Explict baseurl(works correctly)
```
[Article 1]({{site.baseurl}}/{% post_url DATE-article1.md %})
```

The same issue is faced when using images in posts:

- Without Explicit baseurl
```
![Image](/assets/image1.png)
```
>ERROR `/assets/image1.png' not found.

- With Explicit baseurl
```
![Image]({{site.baseurl}}/assets/image1.png)
```

**Comment by ashmaroli** [comment]
@Coder-RG Will you be able *prove* this with a test-repository?

**Comment by Coder-RG** [comment]
@ashmaroli Yes definitely. I am creating my own blog and I faced this issue firsthand. This is why I tried to provide the best explanation for my issue so that it can be recreated. However, if you would like me to demonstrate this, then please do let me know.

**Comment by ashmaroli** [comment]
> However, if you would like me to demonstrate this,

Yes please. If the repo you're already working on is public, you may provide us with the repo URL, else a test repository that I can clone locally and reproduce the issue will suffice.

**Comment by Coder-RG** [comment]
@ashmaroli You may use the repo I am currently working on: [link](https://github.com/Coder-RG/blog/docs)

You can recreate the issue in `2021-11-16-random-agent-tictactoe.markdown`. Just remove `{{site.baseurl}}` mentions and run serve locally. It should give the previously mentioned errors. Do test for both the post_url and the image inclusion URLs.

**Comment by ashmaroli** [comment]
First of all, you're using GitHub Pages.
Therefore, not Jekyll 4.2.

Try using Jekyll 4.2.1 explicitly (comment out gem `github-pages`).

**Comment by Coder-RG** [comment]
@ashmaroli The same issue persists even after commenting out github-pages and using Jekyll explicitly. No error after mentioning `site.baseurl` in the address.
>ERROR `/assets/images/tictactoe.png' not found.

**Comment by ashmaroli** [comment]
`{% post_url ... %}` is meant for posts only. For everything else, please use `{% link ... %}`

**Comment by Coder-RG** [comment]
@ashmaroli I am using post_url for posts only. I was just making you aware that referencing images using the format that I showed above does not include `site.baseurl`.

**Comment by ashmaroli** [comment]
I don't know what more to say.
- The `post_url` and `link` tags prefixes `site.baseurl` from Jekyll 4.0 onward.
- `post_url` is for posts only. For everything else use `link`.
- `post_url` doesn't need the file extension. `link` does.
- Refer https://jekyllrb.com/docs/liquid/tags/ for details..

Closing as we won't be enhancing `post_url` to behave like the `link` tag.

**Comment by Coder-RG** [comment]
I found the issue. My jekyll version was fixed on 3.9.0 even though I had updated it to 4.2.1 using bundle. After I removed Gemfile.lock and ran bundle install, it started working as expected. However, Github pages does does not support 4.x as yet. So I have to keep using 3.9.0 until they update it. Thank you for your prompt replies, by the way!
